Add CSSSelector::parseStandalonePseudoElement()

annevk opened the pull request:
#### d4994b28b489ed0ca40eb28d5399431964076621
<pre>
Add CSSSelector::parseStandalonePseudoElement()
<a href="https://bugs.webkit.org/show_bug.cgi?id=266862">https://bugs.webkit.org/show_bug.cgi?id=266862</a>

Reviewed by Tim Nguyen.

Abstract the logic of non-selector-parser callers of
parsePseudoElement() into a more correct static method.

This means that from now on we require that standalone pseudo-elements
start with &quot;::&quot; (or just &quot;:&quot; for 4 legacy pseudo-elements), which is
aligned with the specification and Firefox.

This also fixes what appears to be a regression in
LocalDOMWindow::getMatchedCSSRules() introduced 272429@main. It
incorrectly assumes to have a CSSSelector::PseudoElement from
which it can obtain a PseudoId.

This also helps <a href="https://bugs.webkit.org/show_bug.cgi?id=243539.">https://bugs.webkit.org/show_bug.cgi?id=243539.</a>

* LayoutTests/fast/css/getComputedStyle/computed-style-properties-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-properties.html:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-with-pseudo-element-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animatable/animate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/KeyframeEffect/target-expected.txt:
* Source/WebCore/animation/DeclarativeAnimationEvent.cpp:
(WebCore::DeclarativeAnimationEvent::DeclarativeAnimationEvent):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setPseudoElement):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::pseudoIdFromString):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
(WebCore::CSSSelector::parseStandalonePseudoElement):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getMatchedCSSRules const):

Canonical link: <a href="https://commits.webkit.org/272499@main">https://commits.webkit.org/272499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77571db45a7fc4153b9a4557fb26c7d12d954cc3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10537 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34347 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7777 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28432 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8893 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28439 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7681 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35694 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28956 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28805 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33963 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7948 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5937 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31822 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28159 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7457 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8617 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -357,6 +357,24 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
     return type;
 }
 
+std::optional<PseudoId> CSSSelector::parseStandalonePseudoElement(StringView input, const CSSSelectorParserContext& context)
+{
+    // FIXME: Tokenize input.
+    if (input.startsWith("::"_s)) {
+        auto pseudoElement = parsePseudoElement(input.substring(2), context);
+        if (!pseudoElement)
+            return std::nullopt;
+        return pseudoId(*pseudoElement);
+    }
+    if (input.startsWith(":"_s)) {
+        auto pseudoClassOrElement = parsePseudoClassAndCompatibilityElementString(input.substring(1));
+        if (!pseudoClassOrElement.compatibilityPseudoElement)
+            return std::nullopt;
+        return pseudoId(*pseudoClassOrElement.compatibilityPseudoElement);
+    }
+    return std::nullopt;
+}
+
```
WebKit has probably a couple of duplicated codes for parsing the values of PseudoElements. ~Last time I checked, there are 3 places in the code. Maybe all of this should be in one place.~ EDIT: It's what you are doing here.

karlcow replied:
The 3 you found and you are trying to fix in this PR.

* [`LocalDomWindow::getMatchedCSSRules`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/page/LocalDOMWindow.cpp#1667-1673)
* [`WebAnimationUtilities::pseudoIdFromString`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/animation/WebAnimationUtilities.cpp#340-364)
* The one I'm [working on](https://github.com/WebKit/WebKit/pull/7468/files#diff-d8731a2ff6dfd96d88c6fc6f49febb2c58020c97094acd2efc17899956913218)  from [`CSSComputedStyleDeclaration::CSSComputedStyleDeclaration`](https://searchfox.org/wubkat/rev/2333404a4a2e271f6ea0f797f792442db05990de/Source/WebCore/css/CSSComputedStyleDeclaration.cpp#50-61)

I'm happy to retire my PR. Yours is more generic. There is still the issue of what to do for the empty list of style.

nt1m commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -1227,17 +1227,16 @@ const String KeyframeEffect::pseudoElement() const
+    // https://drafts.csswg.org/web-animations-1/#dom-keyframeeffect-pseudoelement
+    auto pseudoId = pseudoIdFromString(pseudoElement);
+    if (!pseudoId)
+        return Exception { ExceptionCode::SyntaxError };
```
I know it wasn't there previously, but please add an useful error message as second argument.

nt1m commented on Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
```diff
@@ -51,13 +51,9 @@ CSSComputedStyleDeclaration::CSSComputedStyleDeclaration(Element& element, bool
-        name = name.substring(1);
-    auto pseudoType = CSSSelector::parsePseudoElement(name, CSSSelectorParserContext { element.document() });
-    m_pseudoElementSpecifier = pseudoType ? CSSSelector::pseudoId(*pseudoType) : PseudoId::None;
+    // FIXME: This is quite wrong. https://bugs.webkit.org/show_bug.cgi?id=243539.
```
```suggestion
    // FIXME: This should return a style with initial values if the pseudo-element is invalid (webkit.org/b/243539).
```

karlcow replied:
if anyone is reading that comment in the future. https://github.com/WebKit/WebKit/pull/7468

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -318,7 +318,7 @@ PseudoId CSSSelector::pseudoId(PseudoElement type)
+    if (name.isEmpty())
         return std::nullopt;
 
     auto type = parsePseudoElementString(name);
```
Probably a separate PR, but I kinda wish this was named `parsePseudoElementWithoutContext` or something.

annevk committed:
Add CSSSelector::parseStandalonePseudoElement()
https://bugs.webkit.org/show_bug.cgi?id=266862

Reviewed by Tim Nguyen.

Abstract the logic of non-selector-parser callers of
parsePseudoElement() into a more correct static method.

This means that from now on we require that standalone pseudo-elements
start with "::" (or just ":" for 4 legacy pseudo-elements), which is
aligned with the specification and Firefox.

This also fixes what appears to be a regression in
LocalDOMWindow::getMatchedCSSRules() introduced 272429@main. It
incorrectly assumes to have a CSSSelector::PseudoElement from
which it can obtain a PseudoId.

This also helps https://bugs.webkit.org/show_bug.cgi?id=243539.

* LayoutTests/fast/css/getComputedStyle/computed-style-properties-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-properties.html:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-with-pseudo-element-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-with-pseudo-element.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animatable/animate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/KeyframeEffect/target-expected.txt:
* Source/WebCore/animation/DeclarativeAnimationEvent.cpp:
(WebCore::DeclarativeAnimationEvent::DeclarativeAnimationEvent):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setPseudoElement):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::pseudoIdFromString):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElement):
(WebCore::CSSSelector::parseStandalonePseudoElement):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getMatchedCSSRules const):

Canonical link: https://commits.webkit.org/272499@main
